src/hoisted.jl

```diff
@@ -0,0 +1,84 @@
+# Specialized functions for increased performance when JSON is in-memory
+
+addbyte!(x::Int, c) = x + 1
+addbyte!(x::Vector{UInt8}, c) = push!(x, c)
+
+function parse_string(ps::MemoryParserState)
+    # "Dry Run": find length of string so we can allocate the right amount of
+    # memory from the start.
+    s = ps.s
+    len = parse_string(ps, 0)
+
+    # Now read the string itself
+    ps.s = s
+    ret = Compat.UTF8String(UInt8[])
+    b = sizehint!(ret.data, len)
+    parse_string(ps, b)
+    ret
```

Comment:
This is twice as slow because `String` makes a copy, at least on 0.5.

```julia
julia> const data = repeat(collect(0x01:0x7F); outer=100);

julia> @time str = String(data);
  0.000018 seconds (7 allocations: 12.781 KB)

julia> data[1] = 0x7F
0x7f

julia> str.data[1]
0x01
```